tensorflow/core/common_runtime/zen_layout_pass.cc

```diff
@@ -0,0 +1,1402 @@
+    tf_training_ops.push_back("ApplyAdadelta");
+    tf_training_ops.push_back("ResourceSparseApplyAdadelta");
+    tf_training_ops.push_back("ApplyProximalGradientDescent");
+    tf_training_ops.push_back("SparseApplyProximalGradientDescent");
+    tf_training_ops.push_back("ApplyAdagrad");
+    tf_training_ops.push_back("ApplyAdagradV2");
+    tf_training_ops.push_back("ApplyProximalAdagrad");
+    tf_training_ops.push_back("SparseApplyAdagrad");
+    tf_training_ops.push_back("SparseApplyAdagradV2");
+    tf_training_ops.push_back("SparseApplyProximalAdagrad");
+    tf_training_ops.push_back("ApplyAdagradDA");
+    tf_training_ops.push_back("SparseApplyAdagradDA");
+    tf_training_ops.push_back("ApplyFtrl");
+    tf_training_ops.push_back("ApplyFtrlV2");
+    tf_training_ops.push_back("SparseApplyFtrl");
+    tf_training_ops.push_back("SparseApplyFtrlV2");
+    tf_training_ops.push_back("ApplyMomentum");
+    tf_training_ops.push_back("ApplyKerasMomentum");
+    tf_training_ops.push_back("ApplyAdam");
+    tf_training_ops.push_back("ApplyAdaMax");
+    tf_training_ops.push_back("ApplyRMSProp");
+    tf_training_ops.push_back("ApplyCenteredRMSProp");
+    tf_training_ops.push_back("ApplyAddSign");
+    tf_training_ops.push_back("ApplyPowerSign");
+  }
+
+  // Standard interface to run pass.
+  Status Run(const GraphOptimizationPassOptions &options);
+
+  // Helper function which executes several passes on the graph
+  // for fusion and rewrite
+  //
+  // @input  g - input graph
+  //
+  // Dumps graph before and after rewite
+  //
+  // @return true, if and only if graph is mutated; false otherwise.
+  bool ZenOpRewritePass(std::unique_ptr<Graph> *g);
+
+  // Helper function which rewrites TF-Vanilla ops with Zen ops
+  //
+  // @input  g - input graph
+  //
+  // @return true, if one or more rewrites are successful; false otherwise.
+  bool ZenOpUpdate(std::unique_ptr<Graph> *g);
+
+  // Structure to specify the name of an original node, its new name after
+  // rewrite,  the function to be used to copy attributes for the op, and
+  // the rule (if any) which must hold for rewriting the node
+  typedef struct {
+    string tf_op_name;   // Original name of op of the node in the graph
+    string zen_op_name;  // New name of the op
+    // A function handler to copy attributes from an old node to a new node.
+    std::function<bool(const Node *)> check_validity;
+    // A rule under which to rewrite this node
+    std::function<void(const Node *, NodeBuilder *)> update_zenopattr;
+  } ZenOpRewriteRecord;
+
+ private:
+  // Maintain record about nodes to rewrite
+  std::vector<ZenOpRewriteRecord> zrwdb_;
+
+  // TF training ops list from tensorflow/core/kernels/training_ops.cc
+  std::vector<string> tf_training_ops;
+
+  // Is OpDef::ArgDef a list type? It could be N * T or list(type).
+  // Refer to opdef.proto for details of list type.
+  inline bool ArgIsList(const OpDef::ArgDef &arg) const {
```

Comment:
We have moved the common code from zen_layout_pass.cc to layout_pass_util(.h/.cc) files.
Should we modify [mkl_layout_pass.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/mkl_layout_pass.cc) file as well ?

Reply:
(Sorry I forgot to reply to this one.) 
Thank you for asking! Theoretically, I think we should do it in this PR, but since TF 2.12 branch cut is very soon, I think we should just leave it untouched for now. (If refactoring cause issues for mkl_layout_pass, there won't be enough time to fix them before the branch cut.) I'll handle refactoring mkl_layout_pass after the branch cut. :)

Reply:
Thank you for the clarification!